// @galiciaseguros/system_design/src/index.tsx
import {Card} from './components/card/Card';
import {Button} from './components/buttons/Button';
import {TextArea} from './components/formElements/TextArea';
import {Select} from './components/formElements/Select';
import {Checkbox} from "./components/formElements/Checkbox";
import {Tooltip} from "./components/helpers/Tooltip"
import { TextInput } from './components/formElements/TextInput';
import ThemeProvider  from './hooks/themeProvider';
import {ThemeContext}  from './hooks/themeProvider';
import { Grid } from './components/grid/grid';
import { Box } from './components/grid/box';
import { Breadcrumb } from './components/breadcrumb/Breadcrumb';
import { Tabs } from './components/tab/Tabs';
import {Text} from './components/typography/Text';
import { c } from './colors';
import { Chip } from './components/chip/chip';
import { Table } from './components/table/Table';
export {Table,Chip,c,Text,Grid,Box,Breadcrumb,Tabs,Card,TextInput,Button,ThemeProvider,ThemeContext,TextArea,Tooltip,Checkbox,Select}
