// @galiciaseguros/system_design/src/components/comprobante/comprobante.tsx


import React, { HTMLAttributes } from "react";

import { Page, Text, View, Document, PDFViewer, Image } from '@react-pdf/renderer';
import logo from './../../public/assets/logo.jpg';

export interface Props extends HTMLAttributes<HTMLButtonElement> {
    canvas?: "A4" | "A5",
    title?: string,
    origin?: "OFB" | "ONB",
    tipoOperacion: string,
    details: Array<Detail>,
    conceptos: Array<Detail>,
    legales:string,
}

interface Detail {
    title: string,
    description: string
}

export const Comprobante = ({ legales="",conceptos = [], details = [], canvas = "A4", title = "Document.pdf", origin = "ONB", tipoOperacion = "operación" }: Props) => {

    const logoSrc = logo.src;

    while (details.length < 3) {
        details.push({ title: "Título", description: "Descripción" })
    }
    while (conceptos.length < 3) {
        conceptos.push({ title: "Título", description: "Descripción" })
    }

    const originValue = {
        "ONB": "Online Banking",
        "OFB": "Offline Banking"
    }
    return (
        <PDFViewer style={{ height: "100vh", width: "100%" }}>

            <Document title={title}   >
                <Page size={canvas} style={{ paddingTop: "32px", paddingRight: "40px", paddingBottom: "32px", paddingLeft: "40px" }}   >
                    <View style={{ display: "flex", justifyContent: "space-between", flexDirection: "row" }}>
                        <Image style={{ height: "42px", width: "146px" }} src={`${logoSrc}`}></Image>

                        <View style={{ fontSize: "18px", display: "flex", justifyContent: "flex-end" }}>
                            <Text>{originValue[origin]}</Text>
                        </View>
                    </View>
                    {details.length == 3 &&
                        <View >
                            <View style={{ fontSize: "14px", marginTop: "16px", marginBottom: "8px" }}>
                                <Text >Detalle de {tipoOperacion}</Text>
                            </View>
                            <View style={{ display: "flex", flexDirection: "row", justifyContent: "space-between", fontSize: "12px" }}>
                                <View style={{ fontWeight: "light", fontSize: "12px", display: "flex", flexDirection: "column" }}>
                                    <Text style={{ color: "#6A6A6A" }}>{details[0].title}</Text>
                                    <Text >{details[0].description}</Text>
                                </View>
                                <View style={{ fontWeight: "light", fontSize: "12px", display: "flex", flexDirection: "column" }}>
                                    <Text style={{ color: "#6A6A6A" }}>{details[1].title}</Text>
                                    <Text >{details[1].description}</Text>

                                </View>
                                <View style={{ fontWeight: "light", fontSize: "12px", display: "flex", flexDirection: "column" }}>
                                    <Text style={{ color: "#6A6A6A" }}>{details[2].title}</Text>
                                    <Text >{details[2].description}</Text>

                                </View>
                            </View>
                        </View>

                    }

                    <View style={{ width: "100%", height: "1px", marginTop: "8px", marginBottom: "8px", backgroundColor: "#BEBEBE" }}></View>
                    <View style={{ display: "flex", flexDirection: "column" }}>
                        {conceptos.map((c) => {
                            return (
                                <View style={{ marginTop: "8px", marginBottom: "8px", fontWeight: "light", fontSize: "12px", display: "flex", flexDirection: "column" }}>
                                    <Text style={{ color: "#6A6A6A" }}>{c.title}</Text>
                                    <Text >{c.description}</Text>

                                </View>)
                        }
                        )

                        }
                    </View>
                    <View style={{ width: "100%", height: "1px", marginTop: "8px", marginBottom: "8px", backgroundColor: "#BEBEBE" }}></View>
                    <View style={{fontSize:"12px"}}>
                        <Text style={{color:"#6A6A6A"}}>{legales}</Text>
                    </View>
                </Page>
            </Document>
        </PDFViewer >)
}
