// @galiciaseguros/system_design/src/components/formElements/Switcher.tsx
import React, { HTMLAttributes, useContext, useState } from "react";
import { c } from "../../colors";
import { ThemeContext } from "../../hooks/themeProvider";
import styles from './switcher.module.css'

export interface Props extends HTMLAttributes<HTMLInputElement> {

    disabled?: boolean,
    disabledPosition?: "left" | "right",
    desription?: string,
    textOn?: string,
    textOff?: string,
    initialValue?:boolean
}

export const Switcher = ({ initialValue=false,disabled = false,  textOn = "", textOff = "", desription = "" }: Props) => {

    const [checked,setChecked]=useState(initialValue);
    const {theme}=useContext(ThemeContext)
    
    return (
        <div>
            <div className={styles.switcherInnerContainer}>
                <label className={styles.switch}>
                    <input style={{backgroundColor:checked&&!disabled?c({type:"primary",brand:theme}):""}}  disabled={disabled} className={disabled?styles.inputDisabled:styles.inputEnabled} onChange={(e)=>setChecked(e.target.checked)} checked={checked} type="checkbox" />
                    <span style={{backgroundColor:checked&&!disabled?c({type:"primary",brand:theme}):""}} className={styles.slider}></span>

                </label>
                {(textOn != "" && textOff != "") && <p className={styles.statusText}>{checked?textOn:textOff}</p>}
            </div>
            {desription != "" && <p className={styles.descriptionText}>{desription}</p>}
        </div>
    )
}
