// @galiciaseguros/system_design/src/components/table/TableFooter.tsx
import React, { HTMLAttributes } from "react";
import styles from './tableFooter.module.css'
import { Pagination } from "../pagination/Pagination";
import { useState } from "react";
import { useEffect } from "react";
interface Props extends HTMLAttributes<HTMLTableCellElement> {



    dataType?: 'text' | 'number',
    sortable?: boolean,
    total: number,
    currentPage?: number,
    setResultPerPage: (val: any) => any;
    setPageNumber: (val: any) => any;

}

export const TableFooter = ({ setPageNumber, setResultPerPage, total}: Props) => {


    const [rpp, setRpp] = useState(5)
    const [totalPages, setTotalPages] = useState(1)
    useEffect(() => {
        setResultPerPage(rpp);
        setTotalPages(Math.ceil(total / rpp))
    },
        [rpp])
    useEffect(() => {
        setTotalPages(Math.ceil(total / 5))
    },
        [total])
    return (
        <div className={styles.tableFooterContainer}>
            <div className={styles.paginationContainer}>
                <select
                    style={{ outline: "none", border: "none" }}
                    onChange={(e) => setRpp(parseInt(e.target.value))}
                    defaultValue={rpp}
                >
                    <option value={5}>5</option>
                    <option value={10}>10</option>
                    <option value={20}>20</option>
                    <option value={50}>50</option>
                </select>
                <p style={{ userSelect: "none", margin: "0px" }}>{" "}de{" "}
                    {total}{" "}resultados</p>
           
            </div>
            <Pagination setPage={setPageNumber} totalPages={totalPages}></Pagination>
        </div>
    )
}
