// @galiciaseguros/system_design/src/components/table/TableHeader.tsx
import React, { HTMLAttributes } from "react";
import styles from './tableHeader.module.css'
import { faSort, faChevronUp, faChevronDown } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
interface Props extends HTMLAttributes<HTMLTableCellElement> {


    data?: any,
    dataType?: 'text' | 'number',
    sortable?: boolean,
    sort?: (columnKey: any, order: any) => any;
    columnKey?: string,
    sortState?: 'ASC' | 'DESC' | 'BASE'
}

export const TableHeader = ({ sortState = 'BASE', sort = () => { }, data = "", dataType = "text", columnKey, ...props }: Props) => {

   



    const sorter = {
        'ASC': < FontAwesomeIcon
            onClick={() => { sort(columnKey, 'DESC') }}
            className={styles.sortIcon} icon={faChevronUp}
        ></FontAwesomeIcon>,
        'DESC': < FontAwesomeIcon
            onClick={() => { sort(columnKey, 'ASC') }}
            className={styles.sortIcon} icon={faChevronDown}
        ></FontAwesomeIcon>,
        'BASE': < FontAwesomeIcon
            onClick={() => { sort(columnKey, 'ASC')}}
            className={styles.sortIcon} icon={faSort}
        ></FontAwesomeIcon>,
    }

    const headerFormatter=(data:any)=>{
        return data;
    }

    return (
        <div className={styles.tableHeaderContainer}>
            <p>{headerFormatter(data)}</p>
            {props.sortable &&
                sorter[sortState]
            }
        </div>
    )
}
