// @galiciaseguros/system_design/src/components/progress/ProgressBar.tsx
import React, { FC, HTMLAttributes, useContext } from "react";
import { c } from "../../colors";
import { ThemeContext } from "../../hooks/themeProvider";
import styles from './progressBar.module.css'
export interface Props extends HTMLAttributes<HTMLDivElement> {
    /** Setea el número total de steps*/
    steps?: number,
    /** Setea la posición inicial*/
    position?: number,
    /** Setea el width del progress bar. Puede expresarse en px o %, en caso de que no se indique la unidad se asume px*/
    width?: string ,
    /** Label correspondiente a cada paso del progress bar  */
    labels: string[];

}

export const ProgressBar: FC<Props> = ({ steps = 2, position = 0, width = "100%", labels = ['', ''] }: Props) => {


    const { theme } = useContext(ThemeContext);

    const getInnerWidth=()=>{
        let innerWidth;
        if(typeof(width)=='string'&&width.includes('%')){
            let nWidth=parseFloat(width.slice(0,-1));

            innerWidth=nWidth*(position/steps)+"%"
            return innerWidth;
        }
        else if(typeof(width)=='string'&&width.includes('px')){
            let nWidth=parseFloat(width.slice(0,-2));
            innerWidth=nWidth*(position/steps)+"px"
            return innerWidth;
        }
        else{
            innerWidth=(parseFloat(width)*(position/steps))+"px"
            return innerWidth
        }
    
    }
    const getOuterWidth=()=>{
        if(typeof(width)=='string'&&width.includes('%')){
            return width;
        }
        else if(typeof(width)=='string'&&width.includes('px')){   
            return width;
        }
        else{
       
            return width+"px"
        }
    
    }
    return (
        <div style={{ width: width }} className={styles.progressBarContainer}>
            <div style={{ width: getOuterWidth() }} className={styles.progressBarOuter}>
                <div style={{ width: getInnerWidth(),backgroundColor:c({type:"primary",brand:theme})}} className={styles.progressBarInner}>

                </div>
            </div>
        <p className={styles.progressBarLabel}>{labels[position]}</p>
        </div>
    )
}
