// @galiciaseguros/system_design/src/components/avatar/Badge.tsx
import { faCheck } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import React, { FC, HTMLAttributes } from "react";
import styles from './badge.module.css'

export interface Props extends HTMLAttributes<HTMLDivElement> {

    /**Tamaño del Avatar**/
    size?: 'XXS' | 'XS' | 'SM' | 'MD' | 'LG' | 'XL'
    /**Color de fondo**/
    backgroundColor?: string,
}

export const Badge: FC<Props> = ({ size = "LG", backgroundColor = "#809CB9" }: Props) => {


    const containerProportions = {
        'XXS': 10,
        'XS': 12,
        'SM': 16,
        'MD': 20,
        'LG': 26,
        'XL': 30
    }


    const borderProportions = {
        'XXS': 1,
        'XS': 1,
        'SM': 2,
        'MD': 2,
        'LG': 3,
        'XL': 3
    }


    return (
        <div style={{ width: `${containerProportions[size]}px`, height: `${containerProportions[size]}px` }} className={styles.badgeContainer}>

            <div style={{
                backgroundColor: backgroundColor,
                width: `${containerProportions[size]}px`, height: `${containerProportions[size]}px`,
                borderWidth: `${borderProportions[size]}px`
            }} className={styles.innerBadge}>
                <FontAwesomeIcon
                    icon={faCheck} height={containerProportions[size] / 2} width={containerProportions[size] / 2} color="white" />

            </div>
        </div>
    )
}
