// @galiciaseguros/system_design/src/components/chip/chip.tsx


import React, { HTMLAttributes ,FC} from "react";
import styles from './chip.module.css'
import { Text } from "../typography/Text";

export interface Props extends HTMLAttributes<HTMLDivElement> {
    label: string,
    state: 'Brand' | 'Eminent' | 'Move' | 'Warning' | 'Info' | 'Success' | 'Danger' | 'Grey' | 'Transparent',
    variant: 'light' | 'dark'
}

export const Chip : FC<Props> =  ({ label = "Label", state = 'Info', variant = "light" }: Props) => {


    const bgColor = {
        "Brand": {
            light: "#FFF3EB",
            dark: "#C85000"
        },
        "Eminent": {
            light: "#EFF6FC",
            dark: "#002041"
        },
        "Move": {
            light: "#F3EDFF",
            dark: "#530ADB"
        },
        "Warning": {
            light: "#FDF5E6",
            dark: "#E99D09"
        },
        "Info": {
            light: "#EAF1F8",
            dark: "#2970BC"
        },
        "Success": {
            light: "#E9F3E9",
            dark: "#258825"
        },
        "Danger": {
            light: "#F9E7E7",
            dark: "#C51111"
        },
        "Grey": {
            light: "#E6E6E6",
            dark: "#666666"
        },
        "Transparent": {
            light: "#FFFFFF00",
            dark: "#FFFFFF00"
        },
    }
    const textColor = {
        "Brand": {
            light: "#C85000",
            dark: "#FFFFFF"
        },
        "Eminent": {
            light: "#002041",
            dark: "#FFFFFF"
        },
        "Move": {
            light: "#530ADB",
            dark: "#FFFFFF"
        },
        "Warning": {
            light: "#000000",
            dark: "#000000"
        },
        "Info": {
            light: "#2970BC",
            dark: "#FFFFFF"
        },
        "Success": {
            light: "#258825",
            dark: "#FFFFFF"
        },
        "Danger": {
            light: "#C51111",
            dark: "#FFFFFF"
        },
        "Grey": {
            light: "#666666",
            dark: "#FFFFFF"
        },
        "Transparent": {
            light: "#666666",
            dark: "#666666"
        },
    }
    const borderColor = {
        "Brand": {
            light: "#FFE2CE",
            dark: "#C85000"
        },
        "Eminent": {
            light: "#D8E8F8",
            dark: "#002041"
        },
        "Move": {
            light: "#E2D2FF",
            dark: "#530ADB"
        },
        "Warning": {
            light: "#F6D89D",
            dark: "#E99D09"
        },
        "Info": {
            light: "#BFD4EB",
            dark: "#2970BC"
        },
        "Success": {
            light: "#BEDBBE",
            dark: "#258825"
        },
        "Danger": {
            light: "#EEB8B8",
            dark: "#C51111"
        },
        "Grey": {
            light: "#E6E6E6",
            dark: "#666666"
        },
        "Transparent": {
            light: "#E6E6E6",
            dark: "#E6E6E6"
        },
    }
    return (
        <div className={styles.chipContainer}>
        <div className={styles.chip}  style={{
            backgroundColor:bgColor[state][variant],
            borderColor:borderColor[state][variant]
        }}><Text type="S7" color={textColor[state][variant]} text={label}></Text></div>
        </div>)
}
