// @galiciaseguros/system_design/src/components/card/Card.tsx
import React, { FC, HTMLAttributes, useEffect, useState } from "react";
import { Avatar, AvatarProps } from "../avatar/Avatar";
import { Elipsis } from "../buttons/Elipsis";
import { Text } from "../typography/Text";
import styles from './card.module.css';
//import img1 from './../../public/me.png';
//import Image from "next/image";

export interface Props extends HTMLAttributes<HTMLDivElement> {
    title: string,
    description: string,
    children?: any,
    headerImage?: string,
    footerLinks?: FooterLink[],
    avatarProps?:AvatarProps
}

interface FooterLink {
    label: string,
    href: string
}

export const Card: FC<Props> = ({ footerLinks = [
    { label: "Link1", href: "https://www.google.com.ar" }, { label: "Link2", href: "www.facebook.com" }
], headerImage, children, title = "Título", description = "Description", avatarProps }: Props) => {
    const [mounted, setMounted] = useState(false);
  
    useEffect(() => setMounted(true), []);
    if (!mounted) return null;



    return (
        <div className={styles.cardContainer}>
            {headerImage && <div className={styles.cardHeaderImageContainer}>

            </div>}
            <div className={styles.cardHeaderContainer}>
                <Avatar {...avatarProps} size="MD"></Avatar>
                <div className={styles.titleSubtitleContainer}>
                    <Text type="S4" text={title}></Text>
                    <Text type="S5" color="#6A6A6A" text={description}></Text>
                </div>
                <Elipsis ></Elipsis>
                <div>
                {/*<Image height={200} width={200} alt="IMG!" src={img1}></Image>*/}
                </div>
            </div>
            {children &&
                <div className={styles.cardBodyContainer}>
                    {children}
                </div>}
            {footerLinks.length > 0 && <div className={styles.cardFooterLinks}>
                {footerLinks.map((fl: FooterLink) =>
                    <div className={styles.cardFooterLinkItem}>
                        <Text text={fl.label} href={fl.href} type={"S5"} isLink={true}></Text>
                    </div>
                )}
            </div>}
        </div>
    )
}
